In [1]:
import pandas as pd
import processing

# Filtering effects

In this section we examine the effect of our participants filtering on the various demographics of the sample.

In [2]:
%%time
log = pd.read_csv('./data/log_valid_processed.csv')
log_raw = pd.read_csv('./data/log_valid.csv')
log_raw = processing.get_participants_for_log(log_raw)
participants = pd.read_csv('./data/participants_processed.csv')
participants = participants.set_index('PARTICIPANT_ID')

In [3]:
participants = participants.loc[log_raw.participant_id.unique()]

In [4]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True

In [5]:
participants.is_sampled.value_counts()

False    24862
True      8952
Name: is_sampled, dtype: int64

## Brand and OS

In [6]:
participants.groupby('is_sampled').os.value_counts(normalize=True)

is_sampled  os     
False       Android    0.596493
            iOS        0.403467
            Other      0.000040
True        iOS        0.916890
            Android    0.083110
Name: os, dtype: float64

In [7]:
participants.groupby('is_sampled').browser.value_counts(normalize=True)

is_sampled  browser                   
False       Chrome Mobile                 0.510257
            Mobile Safari                 0.350615
            Facebook                      0.048468
            Chrome Mobile iOS             0.030448
            Chrome Mobile WebView         0.015445
            Samsung Internet              0.015325
            Chrome                        0.010096
            Mobile Safari UI/WKWebView    0.007562
            Opera Mobile                  0.005189
            Amazon Silk                   0.002856
            Pinterest                     0.001810
            UC Browser                    0.000845
            Firefox iOS                   0.000523
            Puffin                        0.000282
            Android                       0.000121
            Coc Coc                       0.000040
            Opera Mini                    0.000040
            Other                         0.000040
            Yandex Browser                0

In [8]:
participants.groupby('is_sampled').device_brand.value_counts(normalize=True).nlargest(20)

is_sampled  device_brand   
True        Apple              0.916890
False       Apple              0.403483
            Generic            0.237802
            Samsung            0.118418
            Generic_Android    0.069144
            LG                 0.039741
            XiaoMi             0.029886
            Huawei             0.025824
True        Samsung            0.022788
            Generic            0.022453
False       vivo               0.015003
True        XiaoMi             0.013963
False       Motorola           0.009292
True        vivo               0.008378
            Generic_Android    0.008155
False       Lenovo             0.007401
            ZTE                0.007200
            Infinix            0.005229
            HTC                0.005189
            Asus               0.004948
Name: device_brand, dtype: float64

**Conclusion:** The valid, post-filtering users are much more heavily oriented towards iOS usage. They also are more likely to use Safari over Chrome

## Demographics

In [9]:
percentiles = [0.05,0.1,0.25,0.5,0.75,0.9,0.95]

In [10]:
participants.groupby('is_sampled').GENDER.value_counts(normalize=True)

is_sampled  GENDER
False       female    0.632290
            male      0.322943
            none      0.044767
True        female    0.697163
            male      0.248436
            none      0.054401
Name: GENDER, dtype: float64

In [11]:
participants.groupby('is_sampled').AGE.describe()

count       mean       std  min   25%   50%   75%   max
is_sampled                                                           
False       24862.0  24.359504  8.775551  6.0  19.0  23.0  28.0  61.0
True         8952.0  22.976318  9.062334  6.0  17.0  21.0  27.0  61.0

**Conclusion:** The valid, post-filtering users have a slightly larger female compostion and are slightly younger.

## Typing

In [12]:
participants.groupby('is_sampled').TYPE_ENGLISH.value_counts(normalize=True)

is_sampled  TYPE_ENGLISH
False       always          0.435811
            usually         0.291562
            sometimes       0.212131
            rarely          0.047297
            never           0.013199
True        always          0.809629
            usually         0.107360
            sometimes       0.058661
            rarely          0.014388
            never           0.009961
Name: TYPE_ENGLISH, dtype: float64

In [13]:
participants.groupby('is_sampled').BROWSER_LANGUAGE.value_counts(normalize=True).nlargest(10)

is_sampled  BROWSER_LANGUAGE          
True        en-us                         0.739723
False       en-us                         0.307135
            en-US,en;q=0.9                0.275762
True        en-gb                         0.083892
False       en-GB,en-US;q=0.9,en;q=0.8    0.074089
True        en-ca                         0.041667
False       en-US                         0.036964
True        en-au                         0.034629
False       en-gb                         0.030689
True        en-US,en;q=0.9                0.024911
Name: BROWSER_LANGUAGE, dtype: float64

In [14]:
participants.groupby('is_sampled').WPM.describe()

count       mean        std       min        25%        50%  \
is_sampled                                                                  
False       24862.0  35.301802  12.963875  0.019106  26.142669  33.572441   
True         8952.0  40.425418  13.421506  0.212161  31.051003  39.433127   

                  75%         max  
is_sampled                         
False       42.658034  154.257192  
True        49.029891  135.426179

**Conclusion:** The valid, post-filtering participants type in English more frequenlty, and type faster. They also have a higher error rate.

## Android users

Given the bias towards iOS users, let's see how each filtering step affects the number of android users.

In [13]:
log_valid = pd.read_csv('./data/log_valid.csv')

In [14]:
# Match the participants
log_valid = processing.get_participants_for_log(log_valid)

In [15]:
%%time
log_valid = processing.log_process(log_valid)

CPU times: user 20min 46s, sys: 1min 55s, total: 22min 41s
Wall time: 23min 34s


In [16]:
%%time
log_valid = processing.mark_entries(log_valid)

CPU times: user 13min 54s, sys: 58.9 s, total: 14min 53s
Wall time: 15min 55s


In [25]:
participants = pd.read_csv('./data/participants_processed.csv')
participants = participants.set_index('PARTICIPANT_ID')

In [26]:
participants.os.value_counts()

Android          19134
iOS              18449
Windows Phone       22
Other                1
Name: os, dtype: int64

#### Valid log

In [29]:
log_valid.shape

(23157825, 12)

In [30]:
log_valid.participant_id.nunique()

33814

In [31]:
participants['is_sampled'] = False
participants.loc[log_valid.participant_id.unique(),'is_sampled'] = True

In [32]:
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android           3560
            iOS                210
            Windows Phone       22
True        iOS              18239
            Android          15574
            Other                1
Name: os, dtype: int64

#### Only forward entries

In [33]:
# Remove all non-forward entries
log = log_valid.loc[log_valid.is_forward].copy()

In [38]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android           3560
            iOS                210
            Windows Phone       22
True        iOS              18239
            Android          15574
            Other                1
Name: os, dtype: int64

#### Remove multichar backend

In [39]:
# Remove participants with the cumulative multichar behaviour
# Action of an entry is single letter AND following two actions in the entry have increasing # of chars
mask = log.key.str.len() == 1
mask &= log.key.shift(-1).str.len() == 2
mask &= log.key.shift(-2).str.len() == 3
mask &= log.entry_id.shift(-1) == log.entry_id
mask &= log.entry_id.shift(-2) == log.entry_id.shift(-1)
participants_invalid = log.loc[mask].participant_id.unique()
log = log.loc[~log.participant_id.isin(participants_invalid)]

In [40]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android          14854
            iOS                254
            Windows Phone       22
True        iOS              18195
            Android           4280
            Other                1
Name: os, dtype: int64

#### Remove swipe users (first phase)

In [42]:
# Remove participants who use swipe
participants_ = processing.get_participants().set_index('PARTICIPANT_ID')
participants_ = participants_.loc[log.participant_id.unique()].copy()
participants_swipe = participants_.loc[participants_.USING_FEATURES.str.contains('swipe')].index
log = log.loc[~log.participant_id.isin(participants_swipe)].copy()
del participants_

In [43]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android          15334
            iOS                935
            Windows Phone       22
True        iOS              17514
            Android           3800
            Other                1
Name: os, dtype: int64

#### Remove swipe users (second phase)

In [47]:
# Remove heavy swype users that managed to get past the earlier swipe filtering stage
# We do this by detect users where swipe was used more than 10 percent of the time 
log = processing.infer_ite_swype(log)
participants_swipe = log.groupby('participant_id').ite.value_counts(
    normalize=True
).unstack()['swype'].sort_values(ascending=False)
participants_swipe = participants_swipe.loc[participants_swipe >= 0.1].index
log = log.loc[~log.participant_id.isin(participants_swipe)].copy()

In [48]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android          15362
            iOS                947
            Windows Phone       22
True        iOS              17502
            Android           3772
            Other                1
Name: os, dtype: int64

#### Remove consistent multichar backend

In [49]:
# Remove participants where every entry before a space is a multichar
# Entry is multicharacter AND is followed by a space AND is zero LD
mask = log.key.str.strip(' ').str.len() > 1
mask &= log.key.shift(-1) == ' '
mask &= log.lev_dist == 0
log['tmp'] = mask

In [50]:
# Find participants who have a large percentage of these 0-LD, multichar entries
participants_multichar = log.groupby('participant_id').tmp.value_counts().unstack()[True].fillna(0)
participants_multichar /= log.groupby(['participant_id','ts_id']).entry_id.nunique().sum(level=0)
participants_multichar = participants_multichar.loc[participants_multichar > 0.2].index
# Remove them
log = log.loc[~log.participant_id.isin(participants_multichar)].copy()

In [51]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android          16446
            iOS                947
            Windows Phone       22
True        iOS              17502
            Android           2688
            Other                1
Name: os, dtype: int64

#### Remove non-native speakers

In [53]:
# Remove non-native speakers
participants_ = processing.get_participants().set_index('PARTICIPANT_ID')
participants_nonnative = participants_.loc[participants_.NATIVE_LANGUAGE != 'en'].index
log = log.loc[~log.participant_id.isin(participants_nonnative)].copy()
del participants_

In [54]:
participants['is_sampled'] = False
participants.loc[log.participant_id.unique(),'is_sampled'] = True
participants.groupby('is_sampled').os.value_counts()

is_sampled  os           
False       Android          17858
            iOS               4549
            Windows Phone       22
            Other                1
True        iOS              13900
            Android           1276
Name: os, dtype: int64

**Conclusion**: Although a few of these filtering steps are baised towards removing Android users. The filtering step that seems to have the largets effect on the OS balance is the removal of the so-called "multicharacter" behaviour. This step changes the balance from around 55% iOS / 45% Android, to around 80% iOS / 20% Android.